ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,108 @@ let host_supports_hvm ~__context host =
+  (** restart stunnel, possibly changing the config file *)
+  val restart : __context:Context.t -> accept:string -> unit
+end = struct
+
+  let cert = !Xapi_globs.server_cert_path
+
+  (** protect ourselves from concurrent writes to files *)
+  module File : sig
+    (** create or update stunnel config file *)
+    val update_config : accept:string -> unit
+    (** calls `generate_ssl_cert` bash script
+      * TODO: don't generate certificates in bash *)
+    val gen_pem_if_not_exist : __context:Context.t -> unit
+  end = struct
+    let xapi_stunnel_conf = "/etc/stunnel/xapi.conf"
+    let m = Mutex.create ()
+    let current_accept = ref None
+
+    let rewrite_xapi_ssl_config_file ~accept =
+      let idle_timeout = match Sys.getenv_opt "STUNNEL_IDLE_TIMEOUT" with
+                         | None -> "; no idle timeout"
+                         | Some x -> Printf.sprintf "TIMEOUTidle = %s" x
+      in
+      let conf_contents = String.concat "\n" [
+          "; autogenerated by helpers.ml"
+        ; "fips = no"
+        ; "pid = /var/run/xapissl.pid"
+        ; "socket = r:TCP_NODELAY=1"
+        ; "socket = a:TCP_NODELAY=1"
+        ; "socket = l:TCP_NODELAY=1"
+        ; "socket = r:SO_KEEPALIVE=1"
+        ; "socket = a:SO_KEEPALIVE=1"
+        ; idle_timeout
+        ; "debug = authpriv.5"
+        ; ""
+        ; "[xapi]"
+        ; Printf.sprintf "accept = %s" accept
+        ; "connect = 80"
+        ; Printf.sprintf "cert = %s" cert
+        ; Printf.sprintf "ciphers = %s" Xcp_const.good_ciphersuites
+        ; "curve = secp384r1"
+        ; "TIMEOUTclose = 1"
+        ; "options = CIPHER_SERVER_PREFERENCE"
+        ; "sslVersion = TLSv1.2"
+        ]
+      in
+      let len = String.length conf_contents in
+      Stdext.Unixext.atomic_write_to_file xapi_stunnel_conf 0o0600
+        (fun fd -> Unix.single_write_substring fd conf_contents 0 len |> ignore)
+
+    let update_config ~accept =
+      Mutex.execute m (fun () ->
+        match !current_accept with
+        | Some current_accept when current_accept = accept -> ()
+        | None | Some _ -> begin
+          current_accept := Some accept;
+          rewrite_xapi_ssl_config_file ~accept
+        end
+      )
+
+    let gen_pem_if_not_exist ~__context =
+      Mutex.execute m (fun () ->
+        if not (Sys.file_exists cert) then begin
+          let hostname = Unix.gethostname () in
+          let cn =
+            if Astring.String.is_prefix ~affix:"localhost" hostname then
+              get_management_ip_addr ~__context
+            else if Astring.String.is_infix ~affix:"." hostname then
+              None
+            else
+              get_management_ip_addr ~__context
+          in
+          call_script "/opt/xensource/libexec/generate_ssl_cert" [cert; Opt.default hostname cn] |> ignore
+        end
+      )
+
+  end
+
+  let systemctl cmd = call_script "/usr/bin/systemctl" [cmd; "stunnel@xapi"]
+  let systemctl_ cmd = systemctl cmd |> ignore
```

Comment:
Do we need both? Another option:
```
ignore @@ call_script "/usr/bin/systemctl" [cmd; "stunnel@xapi"]
```

Reply:
We also have some systemctl in xapi_clustering.ml. If this is more generally useful now feel free to move it (as long as it doesn't introduce a dependency cycle).

Reply:
i'm using `systemctl_ "restart"` and `systemctl "is-enabled"`

i'll look into merging these modules